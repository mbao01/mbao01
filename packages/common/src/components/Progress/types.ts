import * as ProgressPrimitive from "@radix-ui/react-progress";
import { type VariantProps } from "../../libs";
import { getProgressClasses } from "./constants";

export type ProgressProps = Omit<
  React.ComponentPropsWithoutRef<typeof ProgressPrimitive.Root>,
  "asChild"
> &
  VariantProps<typeof getProgressClasses>;
