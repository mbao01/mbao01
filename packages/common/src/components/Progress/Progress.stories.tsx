import type { Meta, StoryFn, StoryObj } from "@storybook/react";
import { Progress } from "./Progress";

const withProgress = (Story: StoryFn) => {
  return (
    <div className="h-80 w-80 flex justify-center items-center">
      <Story />
    </div>
  );
};

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Progress,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
  decorators: [withProgress],
} satisfies Meta<typeof Progress>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {},
};

export const VerticalProgress: Story = {
  args: {
    value: 50,
  },
};

export const HorizontalProgress: Story = {
  args: {
    value: 75,
  },
};

export const NonDecorativeProgress: Story = {
  args: {
    value: 75,
    max: 50,
  },
};

export const ProgressWithCustomLabel: Story = {
  args: {
    value: 75,
    max: 100,
    getValueLabel: (value, max) => `${value} of ${max}`,
  },
};
