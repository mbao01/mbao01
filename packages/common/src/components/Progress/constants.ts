import { cva } from "../../libs";

export const getProgressClasses = cva("relative h-2 w-full overflow-hidden rounded-full", {
  variants: {
    variant: {
      accent: "bg-accent/20",
      base: "bg-base-content/20",
      error: "bg-error/20",
      ghost: "bg-ghost/20",
      info: "bg-info/20",
      link: "bg-link/20",
      neutral: "bg-neutral/20",
      primary: "bg-primary/20",
      secondary: "bg-secondary/20",
      success: "bg-success/20",
      warning: "bg-warning/20",
    },
  },
  defaultVariants: {
    variant: "base",
  },
});

export const getProgressIndicatorClasses = cva("h-full w-full flex-1 transition-all", {
  variants: {
    variant: {
      accent: "bg-accent",
      base: "bg-base-content",
      error: "bg-error",
      ghost: "bg-ghost",
      info: "bg-info",
      link: "bg-link",
      neutral: "bg-neutral",
      primary: "bg-primary",
      secondary: "bg-secondary",
      success: "bg-success",
      warning: "bg-warning",
    },
  },
  defaultVariants: {
    variant: "base",
  },
});
