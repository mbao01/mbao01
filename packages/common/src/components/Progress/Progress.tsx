"use client";

import React from "react";
import * as ProgressPrimitive from "@radix-ui/react-progress";
import { type ProgressProps } from "./types";
import { cn } from "../../utilities";
import { getProgressClasses, getProgressIndicatorClasses } from "./constants";

export const Progress = React.forwardRef<
  React.ElementRef<typeof ProgressPrimitive.Root>,
  ProgressProps
>(({ className, value, max = 100, variant, ...props }, ref) => (
  <ProgressPrimitive.Root
    ref={ref}
    max={max}
    value={value}
    className={cn(getProgressClasses({ variant }), className)}
    {...props}
  >
    <ProgressPrimitive.Indicator
      className={cn(getProgressIndicatorClasses({ variant }))}
      style={{
        transform: `translateX(-${
          (100 * (max - Math.max(Math.min(value ?? 0, max), 0))) / max
        }%)`,
      }}
    />
  </ProgressPrimitive.Root>
));

Progress.displayName = ProgressPrimitive.Root.displayName;
