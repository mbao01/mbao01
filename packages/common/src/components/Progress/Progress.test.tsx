import { render, screen } from "@testing-library/react";
import { describe, expect, it } from "vitest";
import { Progress } from "./Progress";

describe("Progress", () => {
  it("renders a progress bar", () => {
    const { asFragment } = render(<Progress value={77} />);

    const progressbar = screen.getByRole("progressbar");

    expect(progressbar).toHaveAttribute("aria-valuemax", "100");
    expect(progressbar).toHaveAttribute("aria-valuemin", "0");
    expect(progressbar).toHaveAttribute("aria-valuenow", "77");
    expect(progressbar).toHaveAttribute("aria-valuetext", "77%");
    expect(progressbar).toHaveAttribute("data-state", "loading");
    expect(asFragment()).toMatchSnapshot();
  });

  it("has a value label", () => {
    render(<Progress value={50} max={80} getValueLabel={(value, max) => `${value} of ${max}`} />);

    const progressbar = screen.getByRole("progressbar");
    expect(progressbar).toHaveAttribute("aria-valuetext", "50 of 80");
  });

  it.each([
    "accent",
    "base",
    "error",
    "ghost",
    "info",
    "link",
    "neutral",
    "primary",
    "secondary",
    "success",
    "warning",
  ] as const)("has %s variant", (variant) => {
    const label = `${variant} progress bar`;
    const { asFragment } = render(
      <Progress value={50} variant={variant} getValueLabel={() => label} />
    );

    const progressbar = screen.getByRole("progressbar");
    expect(progressbar).toHaveAttribute("aria-valuetext", label);
    expect(asFragment()).toMatchSnapshot();
  });
});
