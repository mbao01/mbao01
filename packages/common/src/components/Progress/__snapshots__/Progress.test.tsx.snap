// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Progress > has accent variant 1`] = `
<DocumentFragment>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="50"
    aria-valuetext="accent progress bar"
    class="relative h-2 w-full overflow-hidden rounded-full bg-accent/20"
    data-max="100"
    data-state="loading"
    data-value="50"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 transition-all bg-accent"
      data-max="100"
      data-state="loading"
      data-value="50"
      style="transform: translateX(-50%);"
    />
  </div>
</DocumentFragment>
`;

exports[`Progress > has base variant 1`] = `
<DocumentFragment>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="50"
    aria-valuetext="base progress bar"
    class="relative h-2 w-full overflow-hidden rounded-full bg-base-content/20"
    data-max="100"
    data-state="loading"
    data-value="50"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 transition-all bg-base-content"
      data-max="100"
      data-state="loading"
      data-value="50"
      style="transform: translateX(-50%);"
    />
  </div>
</DocumentFragment>
`;

exports[`Progress > has error variant 1`] = `
<DocumentFragment>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="50"
    aria-valuetext="error progress bar"
    class="relative h-2 w-full overflow-hidden rounded-full bg-error/20"
    data-max="100"
    data-state="loading"
    data-value="50"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 transition-all bg-error"
      data-max="100"
      data-state="loading"
      data-value="50"
      style="transform: translateX(-50%);"
    />
  </div>
</DocumentFragment>
`;

exports[`Progress > has ghost variant 1`] = `
<DocumentFragment>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="50"
    aria-valuetext="ghost progress bar"
    class="relative h-2 w-full overflow-hidden rounded-full bg-ghost/20"
    data-max="100"
    data-state="loading"
    data-value="50"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 transition-all bg-ghost"
      data-max="100"
      data-state="loading"
      data-value="50"
      style="transform: translateX(-50%);"
    />
  </div>
</DocumentFragment>
`;

exports[`Progress > has info variant 1`] = `
<DocumentFragment>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="50"
    aria-valuetext="info progress bar"
    class="relative h-2 w-full overflow-hidden rounded-full bg-info/20"
    data-max="100"
    data-state="loading"
    data-value="50"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 transition-all bg-info"
      data-max="100"
      data-state="loading"
      data-value="50"
      style="transform: translateX(-50%);"
    />
  </div>
</DocumentFragment>
`;

exports[`Progress > has link variant 1`] = `
<DocumentFragment>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="50"
    aria-valuetext="link progress bar"
    class="relative h-2 w-full overflow-hidden rounded-full bg-link/20"
    data-max="100"
    data-state="loading"
    data-value="50"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 transition-all bg-link"
      data-max="100"
      data-state="loading"
      data-value="50"
      style="transform: translateX(-50%);"
    />
  </div>
</DocumentFragment>
`;

exports[`Progress > has neutral variant 1`] = `
<DocumentFragment>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="50"
    aria-valuetext="neutral progress bar"
    class="relative h-2 w-full overflow-hidden rounded-full bg-neutral/20"
    data-max="100"
    data-state="loading"
    data-value="50"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 transition-all bg-neutral"
      data-max="100"
      data-state="loading"
      data-value="50"
      style="transform: translateX(-50%);"
    />
  </div>
</DocumentFragment>
`;

exports[`Progress > has primary variant 1`] = `
<DocumentFragment>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="50"
    aria-valuetext="primary progress bar"
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
    data-max="100"
    data-state="loading"
    data-value="50"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 transition-all bg-primary"
      data-max="100"
      data-state="loading"
      data-value="50"
      style="transform: translateX(-50%);"
    />
  </div>
</DocumentFragment>
`;

exports[`Progress > has secondary variant 1`] = `
<DocumentFragment>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="50"
    aria-valuetext="secondary progress bar"
    class="relative h-2 w-full overflow-hidden rounded-full bg-secondary/20"
    data-max="100"
    data-state="loading"
    data-value="50"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 transition-all bg-secondary"
      data-max="100"
      data-state="loading"
      data-value="50"
      style="transform: translateX(-50%);"
    />
  </div>
</DocumentFragment>
`;

exports[`Progress > has success variant 1`] = `
<DocumentFragment>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="50"
    aria-valuetext="success progress bar"
    class="relative h-2 w-full overflow-hidden rounded-full bg-success/20"
    data-max="100"
    data-state="loading"
    data-value="50"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 transition-all bg-success"
      data-max="100"
      data-state="loading"
      data-value="50"
      style="transform: translateX(-50%);"
    />
  </div>
</DocumentFragment>
`;

exports[`Progress > has warning variant 1`] = `
<DocumentFragment>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="50"
    aria-valuetext="warning progress bar"
    class="relative h-2 w-full overflow-hidden rounded-full bg-warning/20"
    data-max="100"
    data-state="loading"
    data-value="50"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 transition-all bg-warning"
      data-max="100"
      data-state="loading"
      data-value="50"
      style="transform: translateX(-50%);"
    />
  </div>
</DocumentFragment>
`;

exports[`Progress > renders a progress bar 1`] = `
<DocumentFragment>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="77"
    aria-valuetext="77%"
    class="relative h-2 w-full overflow-hidden rounded-full bg-base-content/20"
    data-max="100"
    data-state="loading"
    data-value="77"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 transition-all bg-base-content"
      data-max="100"
      data-state="loading"
      data-value="77"
      style="transform: translateX(-23%);"
    />
  </div>
</DocumentFragment>
`;
